### PR TITLE
Merge development branch

### PR DESCRIPTION
**Pull request by tcondit**

### CODE REVIEW

**Comment by jamesdehart** [comment]
> No description provided.

FAIL

File: README.md [hunk 1]
```diff
@@ -1,31 +1,48 @@
 # pulumi-practice
 
-### Learning Pulumi by doing. Mostly Python.
+_Learning Pulumi by doing. Mostly Python._
 
-This repo contains infrastructure code for AWS, built with Python-flavored Pulumi. Each subdirectory contains a standalone stack. As an experiment, each of these stacks is a separate release. We'll see if that works, or if not, where it breaks.
+## Learning what now?
 
-## Get ready
+> Pulumi is a platform for building and deploying cloud infrastructure and applications in your favorite language on any cloud.
```
Comment (jamesdehart) [inline README.md]: 🔥

File: README.md [hunk 2]
```diff
@@ -1,31 +1,48 @@
...
-If you skipped that, here's the full list:
+Future possibilities include:
 
+* infrastructure on AWS in TypeScript
+* ready-to-run stacks for Docker, Kubernetes or Serverless (using _Python_ or _TypeScript_)
+* ready-to-run stacks for other cloud providers (using _Python_ or _TypeScript_)
+
+Quick note: Each subdirectory contains a standalone stack. As an experiment, each of these stacks is a separate release. We'll see if that works, or if not, where it breaks.
+
+## Ready ...
```
Comment (jamesdehart) [inline README.md]: 🔥 🔥

File: README.md [hunk 3]
```diff
@@ -1,31 +1,48 @@
...
-  * Note: If you go with option one, you can stop after doing the initial install. Since v3.3, Python has shipped with [venv](https://docs.python.org/3/library/venv.html) which is what we'll use to create and manage virtual environments.
+  * If you go with option one, you can stop after doing the initial install. Since v3.3, Python has shipped with [venv](https://docs.python.org/3/library/venv.html) which is what we'll use to create and manage virtual environments.
 * pip3, but note that it should have been included with Python v3.6 or later. Link provided just in case, but chances are good you won't need it. ([install](https://pip.pypa.io/en/stable/installing/))
 * A Pulumi account ([sign up](https://app.pulumi.com/welcome))
 * The Pulumi CLI ([install](https://pulumi.io/quickstart/))
 * AWS credentials ([configure](https://pulumi.io/quickstart/aws/setup.html))
 * An SSH key pair ([create](https://help.github.com/en/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#generating-a-new-ssh-key))
 
-## Get set
+## Set ...
```
Comment (jamesdehart) [inline README.md]: 🔥 🔥 🔥

File: README.md [hunk 4]
```diff
@@ -192,7 +209,21 @@ Just supply the basename of your private key with neither a path, nor an extensi
 
 * Your key needs to be `chmod 600`
```
Comment (jamesdehart) [inline README.md]: AWS says `400` in this [doc](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AccessingInstancesLinux.html)

File: README.md [hunk 5]
```diff
@@ -192,7 +209,21 @@ Just supply the basename of your private key with neither a path, nor an extensi
...
+* I haven't explained [stack exports](https://pulumi.io/tour/programs-exports.html) yet! They're great! Here's how to export an attribute called `elasticIP`, and make it available to query. The `eip` Resource was defined earlier.
+
+    ```bash
+    pulumi.export('elasticIP', eip.public_ip)
+    ```
+
+* This allows us to make part of the call to `ssh` more general.
+
+    ```bash
+    ➜ ssh -l ec2-user -i ~/Downloads/sl-us-west-2.pem $(pulumi stack output elasticIP)
```
Comment (jamesdehart) [inline README.md]: Love those ->

File: README.md [hunk 6]
```diff
@@ -192,7 +209,21 @@ Just supply the basename of your private key with neither a path, nor an extensi
...
+    ```
+
+* __TIP__ If you're not familiar with this shell syntax, `$()` is a [command substitution](http://www.tldp.org/LDP/abs/html/commandsub.html) operator. Whatever is inside the parentheses is executed in a subshell and the result is passed to the enclosing shell. The leading `$` says to show the result. So we're calling `pulumi stack output elasticIP` in a subshell and passing the result to the enclosing shell. Put it together and we're saying "ssh to whatever IP address the `elasticIP` stack export currently contains".
```
Comment (jamesdehart) [inline README.md]: Nice. +1

File: README.md (hunk 4 above)
Comment (tcondit) [reply README.md]: `0400` is too restricted in my view. Either way (`0400` or `0600`) only the owner can execute the file, so from a security point of view they're both acceptable.

File: README.md (hunk 5 above)
Comment (tcondit) [reply README.md]: They're from `zsh spaceship prompt`.

**Comment by tcondit** [comment]
Pleasure doing business with ya.

**Pull request merged**